design_documents/cairo_deployment_scripts.md

```diff
@@ -0,0 +1,335 @@
+given function should be executetd or not (based on state file)
+  - pros:
+    - elegant solution
+    - could be applied selectively (per function) or globally (to main)
+  - cons:
+    - more complex solution
+    - probably not very suitable for runtime based behaviours (like reading json file) - requires more research
+
+An example pseudo-implementation of this could look like this:
+```rust
+// skip_if_done.rs
+use proc_macro_hack::proc_macro_hack;
+
+#[proc_macro_hack]
+pub use skip_if_done_impl::skip_if_done;
+
+
+// skip_if_done_impl.rs
+use quote::quote;
+use std::env;
+
+fn should_execute() -> bool {
+    // Your logic to check whether the function should be executed goes here
+    true
+}
+
+#[macro_export]
+macro_rules! skip_if_done {
+    ($($tokens:tt)*) => {
+        {
+            if should_execute() {
+                $($tokens)*
+            }
+        }
+    };
+}
+
+// main.rs
+#[path = "skip_if_done.rs"]
+mod skip_if_done;
+
+fn main() {
+    skip_if_done! {
+        dispatcher.send_eth(...);
+    }
+}
+```
+
+Given all this, the most obvious solutions seem to be:
+  - for cast subcommands functions:
+    - adjust them to handle checking the file
+  - for dispatcher functions:
+    - implement a higher order function
+
+Picking those would ensure us that:
+  - syntax is relatively boiler-code free (only dispatcher functions are annotated)
+  - idempotency is achieved
+  - it is quite easy to write a script
+
+### example script
+An example deployment script could look like this:
+
+```cairo
+// we might need to rename account functions to avoid confusion
+use cast::starknet_commands::account::create::create as create_account
+use cast::starknet_commands::account::deploy::deploy as deploy_account
+use cast::{get_provider, get_account};
+use cast::starknet_commands::{declare, deploy, invoke, call};
+(...)
+
+fn make_account(provider: &JsonRpcClient<HttpTransport>) -> Result<SingleOwnerAccount<&'a JsonRpcClient<HttpTransport>, LocalWallet>> {
+  let mut prefunder = get_account("user", &provider)?; // the user that will prefund new account 
+  let user = create_account("user1", &provider);
+  let prefund_account = invoke("0x123", "deposit", [1234, user.address], &prefunder);
+  deploy_account("user1", &provider);
+  get_account("user1", &provider)?
+}
+
+fn main() {
+  ler provider = get_provider("http://127.0.0.1:5050/rpc")?;
+  let user = make_account(provider);
+
+  let declared_contract = declare("mycontract", &user, "/path/to/Scarb.toml");
+  let contract = deploy(&declared_contract.class_hash, [], &user);
+
+  let dispatcher = DispatcherMyContract { contract.contract_address };
+
+  skip_if_done(|| {
+        dispatcher.increase_balance(5);
+    });
+  let get_balance = skip_if_done(|| {
+        dispatcher.get_balance();
+    });
+
+  let called_balance = call(&contract.contract_address, "get_balance", [], "latest"); 
+}
```

Comment:
Yes, this will work in cheatcode-like fashion, in order to allow rust code to be used in cairo. I didn't describe the mechanisms in detail, jsut mentioned it should work like that, because the team already has some experience and good practices developed for this.